Source/ARTRealtimePresence.m

```diff
@@ -622,4 +658,295 @@ - (void)failPendingPresence:(ARTStatus *)status {
+    }
+    [self cleanUpAbsentMembers];
+}
+
+- (void)sync {
+    [self sync:nil];
+}
+
+- (void)sync:(ARTCallback)callback {
+    if (callback) {
+        ARTCallback userCallback = callback;
+        callback = ^(ARTErrorInfo *__nullable error) {
+            dispatch_async(self->_userQueue, ^{
+                userCallback(error);
+            });
+        };
+    }
+
+    switch (_channel.state) {
+        case ARTRealtimeChannelInitialized:
+        case ARTRealtimeChannelDetaching:
+        case ARTRealtimeChannelDetached: {
+            ARTErrorInfo *error = [ARTErrorInfo createWithCode:ARTErrorBadRequest
+                                                       message:@"Unable to sync to channel; not attached."];
+            ARTLogError(self.logger, @"%@", error.message);
+            if (callback) callback(error);
+            return;
+        }
+        default:
+            break;
+    }
+
+    ARTLogVerbose(self.logger, @"R:%p C:%p (%@) requesting a sync operation", _realtime, _channel, _channel.name);
+
+    ARTProtocolMessage *msg = [[ARTProtocolMessage alloc] init];
+    msg.action = ARTProtocolMessageSync;
+    msg.channel = _channel.name;
+    msg.channelSerial = _syncChannelSerial;
+
+    [self startSync];
+    [_realtime send:msg sentCallback:^(ARTErrorInfo *error) {
+        if (error) {
+            ARTLogDebug(self.logger, @"R:%p C:%p (%@) SYNC request failed with %@", self->_realtime, self->_channel, self->_channel.name, error);
+            [self endSync];
+            if (callback) callback(error);
+        }
+        else {
+            ARTLogDebug(self.logger, @"R:%p C:%p (%@) SYNC requested with success", self->_realtime, self->_channel, self->_channel.name);
+            if (callback) callback(nil);
+        }
+    } ackCallback:nil];
+}
+
+- (void)onMessage:(ARTProtocolMessage *)message {
+    int i = 0;
+    for (ARTPresenceMessage *p in message.presence) {
+        ARTPresenceMessage *presence = p;
+        if (presence.data && _dataEncoder) {
+            NSError *decodeError = nil;
+            presence = [p decodeWithEncoder:_dataEncoder error:&decodeError];
+            if (decodeError != nil) {
+                ARTErrorInfo *errorInfo = [ARTErrorInfo wrap:[ARTErrorInfo createWithCode:ARTErrorUnableToDecodeMessage message:decodeError.localizedFailureReason] prepend:@"Failed to decode data: "];
+                ARTLogError(self.logger, @"RT:%p C:%p (%@) %@", _realtime, _channel, _channel.name, errorInfo.message);
+            }
+        }
+
+        if (!presence.timestamp) {
+            presence.timestamp = message.timestamp;
+        }
+
+        if (!presence.id) {
+            presence.id = [NSString stringWithFormat:@"%@:%d", message.id, i];
+        }
+
+        if ([self add:presence]) {
+            [self broadcast:presence];
+        }
+
+        ++i;
+    }
+}
+
+- (void)onSync:(ARTProtocolMessage *)message {
```

Comment:
Can you please refer to https://github.com/ably/ably-go/blob/dd8236fc301eb078803e83efee7615f3ba133341/ably/realtime_presence.go#L276 to check if we are not missing anything here.
We have a separate method to calculate `syncserial` https://github.com/ably/ably-go/blob/dd8236fc301eb078803e83efee7615f3ba133341/ably/realtime_presence.go#L135.

Reply:
Yes, it's similar - you check serials[1] on emptiness, it's done [here](https://github.com/ably/ably-cocoa/blob/913659d50c46a064ef28fbde5e72ea45d09e1d00/Source/ARTRealtimeChannel.m#L676).